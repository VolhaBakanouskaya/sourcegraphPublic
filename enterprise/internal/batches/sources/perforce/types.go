package perforce

import "github.com/sourcegraph/sourcegraph/internal/extsvc/perforce"

// from @varsanojidan, in a comment on the PR:
// You only really need this if the Changelist model doesn't contain information about changelist builds or reviews
// (if perforce has such a concept), if you can get all the info from Changelist directly,
// then you can just use the Changelist type instead.
//
// at some point we will probably want to add info about CLs that we would get from Swarm or elsewhere
// so I'll leave this in here
type AnnotatedChangelist struct {
	*perforce.Changelist
}
