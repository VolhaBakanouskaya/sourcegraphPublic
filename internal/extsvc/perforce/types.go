package perforce

import (
	"time"
)

// TODO: this was just blindly copied from the ADO implementation in order to make things "work" initially
// need to probably switch most of this out for Changelist vernacular

type ChangelistStatus string

type CreatorInfo struct {
	ID          string `json:"id"`
	DisplayName string `json:"displayName"`
	UniqueName  string `json:"uniqueName"`
	URL         string `json:"url"`
	ImageURL    string `json:"imageUrl"`
}

type Changelist struct {
	Depot        Depot            `json:"depot"`
	ID           int              `json:"changelistId"`
	Status       ChangelistStatus `json:"status"` // pending, submitted, shelved
	CreationDate time.Time        `json:"creationDate"`
	CreatedBy    CreatorInfo      `json:"createdBy"`
	Title        string           `json:"title"`
	Description  string           `json:"description"`
	URL          string           `json:"url"`
}

type Repository struct {
	ID         string `json:"id"`
	Name       string `json:"name"`
	CloneURL   string `json:"remoteURL"`
	IsDisabled bool   `json:"isDisabled"`
}
