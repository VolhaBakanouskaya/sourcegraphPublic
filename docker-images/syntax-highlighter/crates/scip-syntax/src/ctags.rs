use std::{io::BufWriter, io::Write, ops::Not, path};

use protobuf::EnumOrUnknown;
use scip::types::{descriptor::Suffix, Descriptor};
use scip_treesitter_languages::parsers::BundledParser;
use serde::{Deserialize, Serialize};

use crate::{get_globals, globals::Scope};

#[derive(Debug)]
pub enum TagKind {
    Function,
    Class,
}

#[derive(Debug)]
pub struct TagEntry {
    pub descriptors: Vec<Descriptor>,
    pub kind: TagKind,
    pub parent: Option<Box<TagEntry>>,

    pub line: usize,
    // pub column: usize,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(tag = "command")]
pub enum Request {
    #[serde(rename = "generate-tags")]
    GenerateTags {
        // command == generate-tags
        filename: String,
        size: usize,
    },
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(tag = "_type")]
pub enum Reply {
    #[serde(rename = "program")]
    Program { name: String, version: String },
    #[serde(rename = "completed")]
    Completed { command: String },
    #[serde(rename = "error")]
    Error { message: String, fatal: bool },
    #[serde(rename = "tag")]
    Tag {
        name: String,
        path: String,
        language: String,
        /// Starts at 1
        line: usize,
        kind: String,
        pattern: String,
        scope: Option<String>,
        #[serde(rename = "scopeKind")]
        scope_kind: Option<String>,
        signature: Option<String>,
        // TODO(SuperAuguste): Any other properties required? Roles? Access?
    },
}

pub fn write_to_buf_writer<T: serde::ser::Serialize, W: std::io::Write>(
    buf_writer: &mut BufWriter<W>,
    val: &T,
) {
    buf_writer
        .write_all(serde_json::to_string(val).unwrap().as_bytes())
        .unwrap();
    buf_writer.write_all("\n".as_bytes()).unwrap();
}

pub fn generate_tags<W: std::io::Write>(
    buf_writer: &mut BufWriter<W>,
    filename: String,
    file_data: &[u8],
) {
    let path = path::Path::new(&filename);

    let parser = match BundledParser::get_parser_from_extension(
        path.extension().unwrap_or_default().to_str().unwrap(),
    ) {
        None => return,
        Some(parser) => parser,
    };

    let (root_scope, _) = match match get_globals(parser, &file_data) {
        None => return,
        Some(res) => res,
    } {
        Err(_) => return,
        Ok(vals) => vals,
    };

    emit_tags_for_scope(
        buf_writer,
        path.file_name().unwrap().to_str().unwrap(),
        vec![],
        &root_scope,
    );
}

fn suffix_to_string(suffix: EnumOrUnknown<Suffix>) -> String {
    return match suffix.enum_value_or_default() {
        // TODO(SuperAuguste): handle more cases + we lose info here, how do handle this?
        Suffix::Namespace => "namespace",
        Suffix::Package => "package",
        Suffix::Method => "method",
        Suffix::Type => "type",
        _ => "variable",
    }
    .to_string();
}

fn emit_tags_for_scope<W: std::io::Write>(
    buf_writer: &mut BufWriter<W>,
    path: &str,
    parent_scopes: Vec<String>,
    scope: &Scope,
) {
    let curr_scopes = {
        let mut curr_scopes = parent_scopes.clone();
        for desc in &scope.descriptors {
            curr_scopes.push(desc.name.clone());
        }
        curr_scopes
    };

    if scope.descriptors.len() > 0 {
        write_to_buf_writer(
            buf_writer,
            &Reply::Tag {
                name: scope
                    .descriptors
                    .iter()
                    .map(|d| d.name.clone())
                    .collect::<Vec<String>>()
                    .join("."),
                path: path.to_string(),
                // TODO(SuperAuguste): Set to correct language (does this even matter?)
                language: "Go".to_string(),
                line: scope.range[0] as usize + 1,
                kind: suffix_to_string(scope.descriptors.last().unwrap().suffix),
                pattern: "/.*/".to_string(),
                scope: parent_scopes
                    .is_empty()
                    .not()
                    .then(|| parent_scopes.join(".")),
                scope_kind: Option::None,
                signature: Option::None,
            },
        );
    }

    for subscope in &scope.children {
        emit_tags_for_scope(buf_writer, path, curr_scopes.clone(), &subscope);
    }

    for global in &scope.globals {
        let mut scope_name = curr_scopes.clone();
        scope_name.extend(
            global
                .descriptors
                .iter()
                .take(global.descriptors.len() - 1)
                .map(|d| d.name.clone()),
        );

        write_to_buf_writer(
            buf_writer,
            &Reply::Tag {
                name: global.descriptors.last().unwrap().name.clone(),
                path: path.to_string(),
                // TODO(SuperAuguste): Set to correct language (does this even matter?)
                language: "Go".to_string(),
                line: global.range[0] as usize + 1,
                kind: suffix_to_string(global.descriptors.last().unwrap().suffix),
                pattern: "/.*/".to_string(),
                scope: scope_name.is_empty().not().then(|| scope_name.join(".")),
                scope_kind: Option::None,
                signature: Option::None,
            },
        );
    }
}
