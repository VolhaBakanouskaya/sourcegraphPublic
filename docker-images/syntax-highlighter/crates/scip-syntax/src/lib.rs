use anyhow::Result;
use scip::types::Occurrence;
use scip_treesitter_languages::parsers::BundledParser;
use tree_sitter::Parser;

pub mod byterange;
pub mod ctags;
pub mod globals;
pub mod languages;
pub mod locals;
pub mod ts_scip;

pub fn get_globals<'a>(
    parser: BundledParser,
    source_bytes: &'a [u8],
) -> Option<Result<(globals::Scope, usize)>> {
    let config = languages::get_tag_configuration(parser)?;
    let mut parser = Parser::new();
    parser.set_language(config.language).unwrap();
    let tree = parser.parse(source_bytes, None).unwrap();
    Some(globals::parse_tree(&config, &tree, source_bytes))
}

pub fn get_locals(parser: BundledParser, source_bytes: &[u8]) -> Option<Result<Vec<Occurrence>>> {
    let mut config = languages::get_local_configuration(parser)?;
    let tree = config.parser.parse(source_bytes, None).unwrap();
    Some(locals::parse_tree(&mut config, &tree, source_bytes))
}

#[cfg(test)]
mod test {
    use std::io::BufWriter;

    use crate::ctags::generate_tags;

    #[test]
    fn test_generate_ctags_go_globals() {
        let mut buffer = vec![0u8; 1024];
        let mut buf_writer = BufWriter::new(&mut buffer);

        generate_tags(
            &mut buf_writer,
            "go-globals.go".to_string(),
            include_bytes!("../testdata/go-globals.go"),
        );
        insta::assert_snapshot!(String::from_utf8_lossy(buf_writer.buffer()));
    }

    #[test]
    fn test_generate_ctags_empty_scope() {
        let mut buffer = vec![0u8; 1024];
        let mut buf_writer = BufWriter::new(&mut buffer);

        generate_tags(
            &mut buf_writer,
            "ctags-empty-scope.rs".to_string(),
            include_bytes!("../testdata/ctags-empty-scope.rs"),
        );
        insta::assert_snapshot!(String::from_utf8_lossy(buf_writer.buffer()));
    }
}
