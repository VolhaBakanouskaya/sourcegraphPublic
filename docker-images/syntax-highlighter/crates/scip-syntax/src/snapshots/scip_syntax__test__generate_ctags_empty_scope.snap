---
source: crates/scip-syntax/src/lib.rs
expression: "String::from_utf8_lossy(buf_writer.buffer())"
---
{"_type":"tag","name":"Arguments","path":"ctags-empty-scope.rs","language":"Go","line":10,"kind":"type","pattern":"/.*/","scope":null,"scopeKind":null,"signature":null}
{"_type":"tag","name":"ParseTiming","path":"ctags-empty-scope.rs","language":"Go","line":15,"kind":"type","pattern":"/.*/","scope":null,"scopeKind":null,"signature":null}
{"_type":"tag","name":"main","path":"ctags-empty-scope.rs","language":"Go","line":28,"kind":"method","pattern":"/.*/","scope":null,"scopeKind":null,"signature":null}
{"_type":"tag","name":"measure_parsing","path":"ctags-empty-scope.rs","language":"Go","line":24,"kind":"method","pattern":"/.*/","scope":null,"scopeKind":null,"signature":null}
{"_type":"tag","name":"parse_files","path":"ctags-empty-scope.rs","language":"Go","line":20,"kind":"method","pattern":"/.*/","scope":null,"scopeKind":null,"signature":null}

